File: forwarder-cc/gateway_sync.cc
#include "gateway_sync.h"

#include <cctype>
#include <cstdlib>
#include <sstream>
#include <unordered_set>

namespace forwarder_cc {
namespace {

std::string Trim(const std::string& s) {
  size_t start = s.find_first_not_of(" \t\r\n");
  if (start == std::string::npos) return "";
  size_t end = s.find_last_not_of(" \t\r\n");
  return s.substr(start, end == std::string::npos ? std::string::npos : end - start + 1);
}

bool HasGatewaySyncOptions(const Config& c) {
  return !c.gateway_location.empty() || !c.gateway_eui.empty() || !c.gateway_eid.empty() ||
         !c.gateway_name.empty() || !c.gateway_radios.empty();
}

int HexVal(char c) {
  if (c >= '0' && c <= '9') return c - '0';
  if (c >= 'a' && c <= 'f') return c - 'a' + 10;
  if (c >= 'A' && c <= 'F') return c - 'A' + 10;
  return -1;
}

}  // namespace

bool ParseGatewayEUI(const std::string& raw, std::vector<uint8_t>& out, std::string& error) {
  std::string s = Trim(raw);
  std::string hex;
  hex.reserve(s.size());
  for (char c : s) {
    if (c == '-' || c == ':') continue;
    hex.push_back(static_cast<char>(std::tolower(static_cast<unsigned char>(c))));
  }
  if (hex.size() != 16) {
    error = "gateway-eui must be 16 hex characters (8 bytes), got " + std::to_string(hex.size()) +
            " chars";
    return false;
  }
  out.resize(8);
  for (size_t i = 0; i < 8; ++i) {
    int hi = HexVal(hex[i * 2]);
    int lo = HexVal(hex[i * 2 + 1]);
    if (hi < 0 || lo < 0) {
      error = "gateway-eui: invalid hex";
      return false;
    }
    out[i] = static_cast<uint8_t>((hi << 4) | lo);
  }
  return true;
}

bool ParseGatewayLocation(const std::string& raw, api::GatewayLocation& out, std::string& error) {
  std::string s = Trim(raw);
  std::vector<std::string> parts;
  {
    std::stringstream ss(s);
    std::string item;
    while (std::getline(ss, item, ',')) parts.push_back(Trim(item));
  }
  if (parts.size() != 2 && parts.size() != 3) {
    error = "gateway-location must be lat,lon or lat,lon,alt";
    return false;
  }
  char* end = nullptr;
  double lat = std::strtod(parts[0].c_str(), &end);
  if (!end || *end != '\0') {
    error = "gateway-location latitude: invalid number";
    return false;
  }
  double lon = std::strtod(parts[1].c_str(), &end);
  if (!end || *end != '\0') {
    error = "gateway-location longitude: invalid number";
    return false;
  }
  if (lat < -90.0 || lat > 90.0) {
    error = "gateway-location latitude out of range: " + parts[0];
    return false;
  }
  if (lon < -180.0 || lon > 180.0) {
    error = "gateway-location longitude out of range: " + parts[1];
    return false;
  }
  out.set_latitude(lat);
  out.set_longitude(lon);
  if (parts.size() == 3) {
    double alt = std::strtod(parts[2].c_str(), &end);
    if (!end || *end != '\0') {
      error = "gateway-location altitude: invalid number";
      return false;
    }
    out.set_altitude(alt);
  }
  return true;
}

bool ParseGatewayRadios(const std::string& raw,
                        std::vector<api::GatewayRadio>& out,
                        std::string& error) {
  out.clear();
  std::unordered_set<int32_t> seen;
  std::stringstream ss(raw);
  std::string entry;
  while (std::getline(ss, entry, ',')) {
    entry = Trim(entry);
    if (entry.empty()) continue;
    std::vector<std::string> parts;
    {
      std::stringstream es(entry);
      std::string p;
      while (std::getline(es, p, ':')) parts.push_back(Trim(p));
    }
    if (parts.size() != 2 && parts.size() != 3) {
      error = "gateway-radios entry \"" + entry +
              "\" must be rf_chain_id:max_tx_power[:tx_sensitivity]";
      return false;
    }
    char* end = nullptr;
    long chain = std::strtol(parts[0].c_str(), &end, 10);
    if (!end || *end != '\0') {
      error = "gateway-radios rf_chain_id: invalid number";
      return false;
    }
    float power = static_cast<float>(std::strtod(parts[1].c_str(), &end));
    if (!end || *end != '\0') {
      error = "gateway-radios max_tx_power: invalid number";
      return false;
    }
    int32_t rf = static_cast<int32_t>(chain);
    if (seen.count(rf)) {
      error = "gateway-radios duplicate rf_chain_id " + std::to_string(rf);
      return false;
    }
    seen.insert(rf);
    api::GatewayRadio radio;
    radio.set_rf_chain(rf);
    radio.set_max_tx_power(power);
    if (parts.size() == 3) {
      float sens = static_cast<float>(std::strtod(parts[2].c_str(), &end));
      if (!end || *end != '\0') {
        error = "gateway-radios tx_sensitivity: invalid number";
        return false;
      }
      radio.set_gain_sensitivity(sens);
    }
    out.push_back(std::move(radio));
  }
  if (out.empty()) {
    error = "gateway-radios is empty";
    return false;
  }
  return true;
}

bool ParseGatewaySyncConfig(const Config& c, ParsedGatewaySync& parsed, std::string& error) {
  parsed = ParsedGatewaySync{};
  error.clear();
  if (!HasGatewaySyncOptions(c)) {
    return false;  // no sync options; not an error
  }

  parsed.has_any = true;
  parsed.gateway_id = Trim(c.gateway_id);
  parsed.external_id = Trim(c.gateway_eid);
  parsed.name = Trim(c.gateway_name);

  if (!c.gateway_eui.empty()) {
    if (!ParseGatewayEUI(c.gateway_eui, parsed.eui, error)) return false;
  }
  if (!c.gateway_location.empty()) {
    if (!ParseGatewayLocation(c.gateway_location, parsed.location, error)) return false;
    parsed.has_location = true;
  }
  if (!c.gateway_radios.empty()) {
    if (!ParseGatewayRadios(c.gateway_radios, parsed.radios, error)) return false;
    parsed.has_radios = true;
  }
  if (parsed.external_id.empty() && parsed.eui.empty()) {
    error =
        "gateway sync requires gateway-eid or gateway-eui when any gateway-* option is set";
    return false;
  }
  return true;
}

api::ReqGatewayUpsert ToUpsertRequest(const ParsedGatewaySync& parsed) {
  api::ReqGatewayUpsert req;
  if (!parsed.gateway_id.empty()) req.set_gateway_id(parsed.gateway_id);
  if (!parsed.external_id.empty()) req.set_external_id(parsed.external_id);
  if (!parsed.eui.empty()) {
    req.set_gateway_eui(parsed.eui.data(), parsed.eui.size());
  }
  if (!parsed.name.empty()) req.set_name(parsed.name);
  if (parsed.has_location) {
    *req.mutable_location() = parsed.location;
  }
  if (parsed.has_radios) {
    auto* radios = req.mutable_radios();
    for (const auto& r : parsed.radios) {
      *radios->add_values() = r;
    }
  }
  return req;
}

}  // namespace forwarder_cc

File: forwarder/gateway_sync.go
package main

import (
	"encoding/hex"
	"fmt"
	"strconv"
	"strings"

	"github.com/kudzutechnologies/analytics/api"
	log "github.com/sirupsen/logrus"
)

// ParsedGatewaySync holds validated gateway synchronization fields from config.
type ParsedGatewaySync struct {
	GatewayID  string
	ExternalID string
	EUI        []byte
	Name       string
	Location   *api.GatewayLocation
	Radios     []*api.GatewayRadio
	HasAny     bool
}

func (c ForwarderConfig) hasGatewaySyncOptions() bool {
	return c.GatewayLocation != "" ||
		c.GatewayEUI != "" ||
		c.GatewayEID != "" ||
		c.GatewayName != "" ||
		c.GatewayRadios != ""
}

// ParseGatewaySyncConfig validates and parses gateway-* sync options.
// Returns nil when no sync options are present.
func ParseGatewaySyncConfig(c ForwarderConfig) (*ParsedGatewaySync, error) {
	if !c.hasGatewaySyncOptions() {
		return nil, nil
	}

	out := &ParsedGatewaySync{
		HasAny:     true,
		GatewayID:  strings.TrimSpace(c.GatewayId),
		ExternalID: strings.TrimSpace(c.GatewayEID),
		Name:       strings.TrimSpace(c.GatewayName),
	}

	if c.GatewayEUI != "" {
		eui, err := parseGatewayEUI(c.GatewayEUI)
		if err != nil {
			return nil, err
		}
		out.EUI = eui
	}

	if c.GatewayLocation != "" {
		loc, err := parseGatewayLocation(c.GatewayLocation)
		if err != nil {
			return nil, err
		}
		out.Location = loc
	}

	if c.GatewayRadios != "" {
		radios, err := parseGatewayRadios(c.GatewayRadios)
		if err != nil {
			return nil, err
		}
		out.Radios = radios
	}

	if out.ExternalID == "" && len(out.EUI) == 0 {
		return nil, fmt.Errorf("gateway sync requires gateway-eid or gateway-eui when any gateway-* option is set")
	}

	return out, nil
}

func parseGatewayEUI(raw string) ([]byte, error) {
	s := strings.TrimSpace(raw)
	s = strings.ReplaceAll(s, "-", "")
	s = strings.ReplaceAll(s, ":", "")
	s = strings.ToLower(s)
	if len(s) != 16 {
		return nil, fmt.Errorf("gateway-eui must be 16 hex characters (8 bytes), got %d chars", len(s))
	}
	b, err := hex.DecodeString(s)
	if err != nil {
		return nil, fmt.Errorf("gateway-eui: %w", err)
	}
	return b, nil
}

func parseGatewayLocation(raw string) (*api.GatewayLocation, error) {
	parts := strings.Split(strings.TrimSpace(raw), ",")
	if len(parts) != 2 && len(parts) != 3 {
		return nil, fmt.Errorf("gateway-location must be lat,lon or lat,lon,alt")
	}
	lat, err := strconv.ParseFloat(strings.TrimSpace(parts[0]), 64)
	if err != nil {
		return nil, fmt.Errorf("gateway-location latitude: %w", err)
	}
	lon, err := strconv.ParseFloat(strings.TrimSpace(parts[1]), 64)
	if err != nil {
		return nil, fmt.Errorf("gateway-location longitude: %w", err)
	}
	if lat < -90 || lat > 90 {
		return nil, fmt.Errorf("gateway-location latitude out of range: %v", lat)
	}
	if lon < -180 || lon > 180 {
		return nil, fmt.Errorf("gateway-location longitude out of range: %v", lon)
	}
	loc := &api.GatewayLocation{
		Latitude:  lat,
		Longitude: lon,
	}
	if len(parts) == 3 {
		alt, err := strconv.ParseFloat(strings.TrimSpace(parts[2]), 64)
		if err != nil {
			return nil, fmt.Errorf("gateway-location altitude: %w", err)
		}
		loc.Altitude = &alt
	}
	return loc, nil
}

func parseGatewayRadios(raw string) ([]*api.GatewayRadio, error) {
	entries := strings.Split(raw, ",")
	seen := map[int32]struct{}{}
	out := make([]*api.GatewayRadio, 0, len(entries))
	for _, entry := range entries {
		entry = strings.TrimSpace(entry)
		if entry == "" {
			continue
		}
		parts := strings.Split(entry, ":")
		if len(parts) != 2 && len(parts) != 3 {
			return nil, fmt.Errorf("gateway-radios entry %q must be rf_chain_id:max_tx_power[:tx_sensitivity]", entry)
		}
		chain, err := strconv.ParseInt(strings.TrimSpace(parts[0]), 10, 32)
		if err != nil {
			return nil, fmt.Errorf("gateway-radios rf_chain_id: %w", err)
		}
		power, err := strconv.ParseFloat(strings.TrimSpace(parts[1]), 32)
		if err != nil {
			return nil, fmt.Errorf("gateway-radios max_tx_power: %w", err)
		}
		rf := int32(chain)
		if _, ok := seen[rf]; ok {
			return nil, fmt.Errorf("gateway-radios duplicate rf_chain_id %d", rf)
		}
		seen[rf] = struct{}{}
		radio := &api.GatewayRadio{
			RfChain:     rf,
			MaxTxPower:  float32(power),
		}
		if len(parts) == 3 {
			sens, err := strconv.ParseFloat(strings.TrimSpace(parts[2]), 32)
			if err != nil {
				return nil, fmt.Errorf("gateway-radios tx_sensitivity: %w", err)
			}
			v := float32(sens)
			radio.GainSensitivity = &v
		}
		out = append(out, radio)
	}
	if len(out) == 0 {
		return nil, fmt.Errorf("gateway-radios is empty")
	}
	return out, nil
}

func (p *ParsedGatewaySync) toUpsertRequest() *api.ReqGatewayUpsert {
	req := &api.ReqGatewayUpsert{}
	if p.GatewayID != "" {
		id := p.GatewayID
		req.GatewayId = &id
	}
	if p.ExternalID != "" {
		eid := p.ExternalID
		req.ExternalId = &eid
	}
	if len(p.EUI) > 0 {
		req.GatewayEui = p.EUI
	}
	if p.Name != "" {
		name := p.Name
		req.Name = &name
	}
	if p.Location != nil {
		req.Location = p.Location
	}
	if p.Radios != nil {
		req.Radios = &api.GatewayRadios{Values: p.Radios}
	}
	return req
}

// scheduleGatewayUpsert runs GatewayUpsert in a detached goroutine so proxying
// and the analytics flush loop are never blocked by the sync RPC.
func (f *AnalyticsForwarder) scheduleGatewayUpsert(parsed *ParsedGatewaySync) {
	if parsed == nil || !parsed.HasAny {
		return
	}
	req := parsed.toUpsertRequest()
	go func() {
		resp, err := f.client.GatewayUpsert(req)
		if err != nil {
			log.Warnf("Gateway upsert failed: %v", err)
			return
		}
		if resp != nil && resp.Applied {
			log.Infof("Gateway upsert applied (id=%s)", resp.GatewayId)
		} else if resp != nil {
			log.Infof("Gateway upsert acknowledged without mutation (id=%s)", resp.GatewayId)
		} else {
			log.Infof("Gateway upsert acknowledged without mutation")
		}
	}()
}
